src/common/async/async_call.h

```diff
@@ -0,0 +1,215 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 IBM
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+/// \file common/async/async_call.h
+///
+/// \brief Wait for functions to run on arbitrary executors
+
+#include <concepts>
+#include <type_traits>
+
+#include <boost/asio/compose.hpp>
+#include <boost/asio/defer.hpp>
+#include <boost/asio/dispatch.hpp>
+#include <boost/asio/post.hpp>
+
+namespace ceph::async {
+/// \brief Dispatch a function on another executor and wait for it to
+/// finish
+///
+/// Useful with strands and similar
+///
+/// \param executor The executor on which to call the function
+/// \param f The function to call
+/// \param token Boost.Asio completion token
+/// \param ioe Parameter pack of executors and I/O Objects to keep live
+///
+/// \return The return value of `f` in a way appropriate to the
+/// completion token. See Boost.Asio documentation.
+template<std::invocable<> F,
+	 boost::asio::completion_token_for<
+	   void(std::invoke_result_t<F>)> CompletionToken,
+	 typename ...IOE>
+auto async_dispatch(auto executor, F&& f, CompletionToken&& token, IOE&& ...ioe)
+{
+  namespace asio = boost::asio;
+  return asio::async_compose<
+    CompletionToken,
+    void(std::invoke_result_t<F>)>(
+      [&executor, f = std::move(f)] (auto& self) mutable {
+	asio::dispatch(
+	  executor,
+	  [f = std::move(f), self = std::move(self)]() mutable {
+	    self.complete(std::invoke(f));
+	  });
+      }, token, executor, std::forward<IOE>(ioe)...);
+}
+
+/// \brief Dispatch a function on another executor and wait for it to
+/// finish
+///
+/// Useful with strands and similar
+///
+/// \param executor The executor on which to call the function
+/// \param f The function to call
+/// \param token Boost.Asio completion token
+/// \param ioe Parameter pack of executors and I/O Objects to keep live
+///
+/// \return The return value of `f` in a way appropriate to the
+/// completion token. See Boost.Asio documentation.
+template<std::invocable<> F,
+	 boost::asio::completion_token_for<void()> CompletionToken,
+	 typename ...IOE,
+	 typename = std::enable_if_t<std::is_void_v<std::invoke_result_t<F>>>>
+auto async_dispatch(auto executor, F&& f, CompletionToken&& token, IOE&& ...ioe)
+{
+  namespace asio = boost::asio;
+  return asio::async_compose<
+    CompletionToken, void()>(
+      [&executor, f = std::move(f)] (auto& self) mutable {
+	asio::dispatch(
+	  executor,
+	  [f = std::move(f), self = std::move(self)]() mutable {
+	    std::invoke(f);
+	    self.complete();
```

Comment:
I'm trying this:

```
template<std::invocable<> F,
         boost::asio::completion_token_for<
           void(std::invoke_result_t<F>)> CompletionToken,
         typename ...IOE>
auto async_dispatch(boost::asio::execution::executor auto executor,
                    F&& f, CompletionToken&& token, IOE&& ...ioe)
{
  namespace asio = boost::asio;
  auto ex2 = asio::get_associated_executor(token, executor);
  return asio::async_compose<CompletionToken, void(std::invoke_result_t<F>)>(
      [&executor, &ex2, f = std::move(f)](auto &self) mutable {
        asio::dispatch(executor, [f = std::move(f), &ex2,
                                  self = std::move(self)]() mutable {
          auto r = std::invoke(f);
          asio::dispatch(ex2,
                         [r = std::move(r), self = std::move(self)]() mutable {
                           self.complete(std::move(r));
                         });
        });
      },
      token, executor, ex2, std::forward<IOE>(ioe)...);
}
```

But I get a crash in the `asio::dispatch(ex2,` specifically because ` impl_ = std::shared_ptr<boost::asio::detail::strand_executor_service::strand_impl> (empty) = {get() = 0x0}`

(If I don't capture `executor` by reference I get a crash on trying to dispatch to it, which is very surprising to me as it's not how I thought executors were supposed to work.)

Any idea what I might be doing wrong?

Reply:
> (If I don't capture `executor` by reference I get a crash on trying to dispatch to it, which is very surprising to me as it's not how I thought executors were supposed to work.)

passing references to stack variables `executor` and `ex2` may be safe for coroutines, but not for CompletionTokens in general

what executors are involved in this crash? is it possible that ex2's execution context runs out of work and goes out of scope?

Reply:
It shouldn't be able to go out of scope, since I pass it to `async_compose` as one of the Executors-or-I/O-Objects to keep live.

The call is https://github.com/ceph/ceph/blob/876025d14a0fa4cd02c8a5e607c91423e5a56501/src/test/common/test_async_call.cc#L111

The supplied `executor` is a strand and the completion token is `use_awaitable` so I think it should be the executor from the io_context.

Reply:
tried building that test case, but i'm having trouble compiling your co_spawn() fork against bundled boost 1.82:
```
In file included from /home/cbodley/ceph/src/common/async/async_yield.h:30,
                 from /home/cbodley/ceph/src/test/common/test_async_call.cc:16:
/home/cbodley/ceph/src/common/async/co_spawn.hpp:545:28: error: ‘allocate_shared’ has not been declared in ‘boost::asio::detail’
  545 | using boost::asio::detail::allocate_shared;
      |                            ^~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```
i really hope we can find a way to avoid forking co_spawn(), now that we're finally able to drop the fork of spawn()

Reply:
i changed some things back to asio::co_spawn() and am reproducing the crashes you mentioned

Reply:
i see AsyncDispatch.AsyncCall passing with this change:
```diff
   namespace asio = boost::asio;
-  auto ex2 = asio::get_associated_executor(token, executor);
   return asio::async_compose<CompletionToken, void(std::invoke_result_t<F>)>(
-      [&executor, &ex2, f = std::move(f)](auto &self) mutable {
+      [&executor, f = std::move(f)](auto &self) mutable {
         asio::dispatch(executor, [f = std::move(f), &ex2,
                                   self = std::move(self)]() mutable {
           auto r = std::invoke(f);
+          auto ex2 = self.get_executor(); // dispatch back to the completion handler's executor
           asio::dispatch(ex2,
```

Reply:
i still don't understand why it crashes when you capture `executor` by value. capturing by reference is probably ok in most cases, but may be problematic with `asio::deferred`:
```c++
auto init = async_dispatch(ex, func, asio::deferred);

// deferred initiation calls the outer lambda which captured a reference to async_dispatch()'s stack
co_await result = init(asio::use_awaitable);
```

Reply:
ok, i get it now:
```diff
       [executor, f = std::move(f)](auto &self) mutable {
-        asio::dispatch(executor, [f = std::move(f),
-                                  self = std::move(self)]() mutable {
+        auto ex = executor; // copy executor out of the lambda before it moves
+        asio::dispatch(ex, [f = std::move(f), self = std::move(self)]() mutable {
```
with `async_compose()`, the lambda itself is stored in the `composed_op` that we refer to as `auto& self`. as part of this call to `dispatch()`, we move `self` into the inner lambda capture, but refer to `executor` from the outer lambda capture in the same statement. the crash shows a null _impl pointer there because it's a moved-from handle

Reply:
This gets everything working perfectly, thank you. I'm going to add some tests involving deferred.